sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/TemporalJoin.java

```diff
@@ -0,0 +1,491 @@
+        Duration asOfOffset,
+        Duration within,
+        Duration maximumJoinPeriod) {
+      this.mainValueCoder = mainValueCoder;
+      this.rhsValueCoder = rhsValueCoder;
+      this.mainItems = StateSpecs.orderedList(mainValueCoder);
+      this.rhsItems = StateSpecs.orderedList(rhsValueCoder);
+      this.outputUnjoinedElements = outputUnjoinedElements;
+      this.asOfOffset = asOfOffset;
+      this.within = within;
+      this.maximumJoinPeriod = maximumJoinPeriod;
+    }
+
+    private static Instant bucketTimestamp(Instant ts) {
+      long bucketIndex = ts.getMillis() / TIMER_BUCKET;
+      return Instant.ofEpochMilli(bucketIndex * TIMER_BUCKET);
+    }
+
+    @ProcessElement
+    public void process(
+        @Element KV<K, RawUnionValue> element,
+        @Timestamp Instant ts,
+        @StateId("mainItems") OrderedListState<MainT> mainItems,
+        @StateId("rhsItems") OrderedListState<AgainstT> rhsItems,
+        @TimerFamily("joinTimers") TimerMap joinTimers,
+        @TimerFamily("cleanupTimers") TimerMap cleanupTimers) {
+      RawUnionValue value = element.getValue();
+      Instant bucketStart = bucketTimestamp(ts);
+      Instant bucketEnd = bucketStart.plus(TIMER_BUCKET);
+      switch (value.getUnionTag()) {
+        case MAIN_TAG:
+          mainItems.add(TimestampedValue.of((MainT) value.getValue(), ts));
+          joinTimers
+              .get(Long.toString(bucketEnd.getMillis()))
+              .withOutputTimestamp(bucketStart)
+              .set(bucketEnd);
+          break;
+        case AGAINST_TAG:
+          rhsItems.add(TimestampedValue.of((AgainstT) value.getValue(), ts));
+          if (maximumJoinPeriod != null) {
+            // TODO(reuvenlax): Add allowedLateness once supported.
+            Instant cleanupTime = bucketEnd.plus(maximumJoinPeriod);
+            cleanupTimers.get(Long.toString(bucketEnd.getMillis())).set(cleanupTime);
+          }
+          // TODO(reuvenlax): Once allowedLateness is supported, we might need to reset the timer.
+          // Idea of how this might be implemented:
+          // store a sparse histogram mainVaue item timestamps in state. Also store the maximum
+          // timer firing timestamp
+          // in state. If a rhs item shows up with a timestamp < the maximum timer firing, then
+          // it is late. In this
+          // case walk across the histogram buckets that are > the rhs timestamp, and set those
+          // timers to fire again.
+          // We would also have to move garbage collection of the main items out of the regular
+          // timer and into a separate
+          // TimerMap, which fires only after bucket-end + allowed lateness.
+          // For performance, we skip all the above state if allowedLateness == 0.
+          break;
+        default:
+          throw new RuntimeException("Unexpected union tag " + value.getUnionTag());
+      }
+    }
+
+    @OnTimerFamily("joinTimers")
+    public void onJoinTimer(
```

Comment:
If we dont provide an option for a TimestampedValue, the user will lose the information about the timestamp that was actually matched I think? As the output will have pcollection timestamp of timer event time.